### DISCUSSION TITLE
What's the major differences between flagd vs go-feature-flag

### DISCUSSION THREAD

**Discussion by yzx396** [thomaspoignant/go-feature-flag#740] [Q&A] [answered]
There is a default implementation in Openfeature org so call `flagd`, what are the major differences between the two? @thomaspoignant

**Comment by thomaspoignant** [comment, marked as answer]
FlagD and GO Feature Flag are two open-source solutions designed to help developers experiment with feature flagging in a simple way. Both solutions share the common goal of providing a way to control feature flags and experiment with new features, with the aim of minimizing the risk of bugs and other issues. However, while they have a lot in common, there are also some differences between the two solutions.
The main difference between FlagD and GO Feature Flag is their ecosystem and versatility.

FlagD is a feature flagging solution specifically designed for use in Kubernetes. It is built to work seamlessly with Kubernetes and its associated technologies. Therefore, if you are using Kubernetes, FlagD may be a better fit for you as it has been created with this ecosystem in mind.

GO Feature Flag, on the other hand, is a more versatile solution that can work with a wide range of configuration locations outside of Kubernetes, such as AWS S3, Google Cloud, GitHub, Gitlab, etc... It can also work with configmaps inside Kubernetes. 
GO Feature Flag can be used as a GO module directly or with the relay proxy with compatibility to OpenFeature and offer providers with several languages such as GO, Java, .NET, JavaScript, and Python is on its way.

Moreover, GO Feature Flag has more advanced rollout capabilities with the support of experimentation, progressive delivery, schedule changes, etc... This means that you can use GO Feature Flag to experiment with new features and roll them out to users gradually, helping to minimize the risk of bugs and other issues.

In summary, FlagD is more focused on the Kubernetes ecosystem, while GO Feature Flag is a more versatile and feature-rich solution that can be used in a wider range of environments.

@yangzhaox I hope it gives you the information you were looking for?

**Reply by yzx396** [reply]
I did a quick comparison between the two projects, GOFF Relay Proxy (maybe a better name needed) is the equivalent of FlagD. In general GOFF's design & code is simpler, easier to understand, higher test coverage, and the editor, more rollout strategies and better documentation are big plus. One difference is the two projects are using different rule engine for targeting, FlagD is using https://jsonlogic.com/ and GOFF is using https://github.com/nikunjy/rules. Personally I feel jsonlogic is more expressive and a bette choice. your thoughts? @thomaspoignant

**Reply by thomaspoignant** [reply]
@yangzhaox thanks for your feedback and your detailed points for GO Feature Flag.
Yes, I tried since day one to have the code as simple as possible in the project, I am glad that you fill that it is easier for you to understand it.

About `jsonlogic` I am not a big fan of it because I found it verbose, but it can be straightforward to be compatible with it.
In the core of GO Feature Flag the only place where we are using `nikunjy/rules` is this [line](https://github.com/thomaspoignant/go-feature-flag/blob/46828e84d41904733f9a1bfce553cd55081b11d4/internal/flag/rule.go#LL47C50-L47C56) so it is easy to support both formats.
If you feel that it can help the project to be compatible with it, please open an [issue](https://github.com/thomaspoignant/go-feature-flag/issues/new?assignees=&labels=enhancement%2Cneeds-triage&projects=&template=feature.yaml&title=%28feature%29+%3Ctitle%3E) and I can have a look at it. 

About the naming of the `relay proxy` I am not 100% sure either about this name.
But I had no better idea, so if you can explain me what you don't like with this name it may help me to find a better one 😉

**Reply by yzx396** [reply]
I have an use case to define rule like:
companyUUID in [ 
  "4f7443e9-2f64-433d-8472-008550ca0bb9", 
  "239a936b-e33b-4040-8893-1fea7ad8ad88",
  ...,
  "7afa0655-59d1-4a3a-b422-f9f35d80ac0c"
]
It would be hard to read and update if this rule is defined as a `query` string

As to the name relay proxy, in fact it is a generic feature flag service which allows users to use different programming languages and sdks. I understand GOFF starts as a GO module and then evolves into a service but the naming Go Feature Flag Relay Proxy may give people a wrong impression that it works only in Go.

**Reply by thomaspoignant** [reply]
I am not sure to understand why `jsonlogic` will be better in such case?

I've tested your example in **GO Feature Flag** and we can do something like the example below, I still find it explicit but it is maybe because I am used to it.

```yaml
example-flag:
  variations:
    default_var: false
    false_var: false
    true_var: true
  targeting:
    - query: |
        companyUUID IN ["4f7443e9-2f64-433d-8472-008550ca0bb9",
        "239a936b-e33b-4040-8893-1fea7ad8ad88",
        "7afa0655-59d1-4a3a-b422-f9f35d80ac0c"]
      variation: true_var
  defaultRule:
    percentage:
      false_var: 100
```

**Reply by yzx396** [reply]
Oh this looks great. I was thinking how to do this in json format. 

About jsonlogic, the possible advantage is if one day you want to evaluate the targeting rule in frontend, it has javascript implementation.

**Reply by yzx396** [reply]
Given the above use case, how to use GOFF editor to do that? I feel a single-line input for query is not an easy way to define such a multiple-line rule. And YAML format is better at readability than JSON in this case.

**Reply by yzx396** [reply]
I have an idea to enhance GOFF editor: in fact it can be turned into a GOFF playground where not only a rule can be defined but also it can be tested with manually entered user context data without attaching to GOFF relay proxy. The key to enable this is the rule can be evaluated purely in JS on the frontend. What do you think?

**Reply by thomaspoignant** [reply]
@yangzhaox sorry for the late reply be this is actually a lot of good things here.
The editor was released as an helper, yes it is probably not perfect for now and can be improved a lot.

Since this is not my biggest expertise I haven't spend that much time on it, and I would love to have contributors to help me with it. If by any chance you know someone that can help with that, it would be awesome 👍

**Reply by thomaspoignant** [reply]
@yangzhaox can I ask you to create some issues with your ideas?
I think it will be great to track them.
